specification/communication/data-plane/Microsoft.CommunicationServicesCommon/stable/2021-03-07/common.json

```diff
@@ -0,0 +1,154 @@
+  "info": {
+    "title": "Communication",
+    "description": "Communication Services API common types.",
+    "version": "2021-03-07"
+  },
+  "paths": {},
+  "definitions": {
+    "CommunicationErrorResponse": {
+      "description": "The Communication Services error.",
+      "type": "object",
+      "required": [
+        "error"
+      ],
+      "properties": {
+        "error": {
+          "description": "The Communication Services error.",
+          "$ref": "#/definitions/CommunicationError"
+        }
+      }
+    },
+    "CommunicationError": {
+      "description": "The Communication Services error.",
+      "type": "object",
+      "required": [
+        "code",
+        "message"
+      ],
+      "properties": {
+        "code": {
+          "type": "string",
+          "description": "The error code."
+        },
+        "message": {
+          "type": "string",
+          "description": "The error message."
+        },
+        "target": {
+          "type": "string",
+          "readOnly": true,
+          "description": "The error target."
+        },
+        "details": {
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/CommunicationError"
+          },
+          "readOnly": true,
+          "description": "Further details about specific errors that led to this error."
+        },
+        "innererror": {
+          "x-ms-client-name": "innerError",
+          "readOnly": true,
+          "$ref": "#/definitions/CommunicationError",
+          "description": "The inner error if any."
+        }
+      }
+    },
+    "CommunicationCloudEnvironmentModel": {
+      "description": "The cloud that the identifier belongs to.",
+      "type": "string",
+      "enum": [
+        "public",
+        "dod",
+        "gcch"
+      ],
+      "x-ms-enum": {
+        "name": "CommunicationCloudEnvironmentModel",
+        "modelAsString": true
+      }
+    },
+    "CommunicationIdentifierModel": {
+      "description": "Identifies a participant in Azure Communication services. A participant is, for example, a phone number or an Azure communication user. This model must be interpreted as a union: Apart from rawId, at most one further property may be set.",
+      "type": "object",
+      "properties": {
+        "rawId": {
+          "type": "string",
+          "description": "Raw Id of the identifier. Optional in requests, required in responses."
+        },
+        "communicationUser": {
+          "$ref": "#/definitions/CommunicationUserIdentifierModel",
+          "description": "The communication user."
+        },
+        "phoneNumber": {
+          "$ref": "#/definitions/PhoneNumberIdentifierModel",
+          "description": "The phone number."
+        },
+        "microsoftTeamsUser": {
+          "$ref": "#/definitions/MicrosoftTeamsUserIdentifierModel",
+          "description": "The Microsoft Teams user."
+        }
+      }
+    },
+    "CommunicationUserIdentifierModel": {
```

Comment:
The problem with eliminating explicit "kind" enum and having "optional" full objects, would result customers processing the events as :

if (eventData.CommunicationUser != null)
{
}
else if (eventData.PhoneNumber != null)
{
}
else if (eventData.MicrosoftTeamsUser != null)
{
}
etc.

Not a clean switch case. Also it will be difficult distinguish between an error case(all user representation objects are null) and new user object(which can be detected by default label in a switch case).

So would it be beneficial to include "kind" along with the specific objects?

Reply:
having `kind` helps with deserialziation, but for preparing requests, it is redundant information. Also we were thinking other violation could happen so we need to check the properties anyways. Eg, someone could set multiple properties. We can offer serialization/deserialization in our code to help the developers. Eg: https://github.com/Azure/azure-sdk-for-net/pull/18389/